sklearn/decomposition/_pca.py

```diff
@@ -523,33 +557,88 @@ def _fit_full(self, X, n_components):
-            # but that would make this code rely on numpy's SVD instead of
-            # scipy's. It's not 100% clear whether they use the same LAPACK
-            # solver by default though (assuming both are built against the
-            # same BLAS).
-            U, S, Vt = linalg.svd(X, full_matrices=False)
+        if isinstance(self.mean_, np.matrix):
+            # This can happen when X is a scipy sparse matrix.
+            self.mean_ = np.asarray(self.mean_).ravel()
+
+        if self._fit_svd_solver == "full":
+            X_centered = xp.asarray(X, copy=True) if self.copy else X
+            X_centered -= self.mean_
+            x_is_centered = not self.copy
+
+            if not is_array_api_compliant:
+                # Use scipy.linalg with NumPy/SciPy inputs for the sake of not
+                # introducing unanticipated behavior changes. In the long run we
+                # could instead decide to always use xp.linalg.svd for all inputs,
+                # but that would make this code rely on numpy's SVD instead of
+                # scipy's. It's not 100% clear whether they use the same LAPACK
+                # solver by default though (assuming both are built against the
+                # same BLAS).
+                U, S, Vt = linalg.svd(X_centered, full_matrices=False)
+            else:
+                U, S, Vt = xp.linalg.svd(X_centered, full_matrices=False)
+            explained_variance_ = (S**2) / (n_samples - 1)
+
         else:
-            U, S, Vt = xp.linalg.svd(X, full_matrices=False)
+            assert self._fit_svd_solver == "covariance_eigh"
+            # In the following, we center the covariance matrix C a-posteriori
+            # (without centering the data X first) to avoid an unnecessary copy
+            # of X. Note that the mean_ attribute is still needed to center
+            # test data in the transform method.
+            #
+            # Note: at the time of writing, `xp.cov` does not exist in the
+            # Array API standard:
+            # https://github.com/data-apis/array-api/issues/43
+            #
+            # Besides, using `numpy.cov`, as of numpy 1.26.0, would not be
+            # memory efficient for our use case when `n_samples >> n_features`:
+            # `numpy.cov` centers a copy of the data before computing the
+            # matrix product instead of subtracting a small `(n_features,
+            # n_features)` square matrix, a posteriori, as we do below.
+            x_is_centered = False
+            C = X.T @ X
+            C -= (
+                n_samples
+                * xp.reshape(self.mean_, (-1, 1))
+                * xp.reshape(self.mean_, (1, -1))
+            )
+            C /= n_samples - 1
+            eigenvals, eigenvecs = xp.linalg.eigh(C)
+
+            # The following can happen when X is a scipy sparse matrix.
+            if isinstance(eigenvals, np.matrix):
```

Comment:
I think we should not use `np.matrix` anywhere.

Reply:
The point of this line is precisely to get rid of such `np.matrix` occurrences that typically show up when using accepting `scipy.sparse.csr_matrix` and the like as inputs. `scipy.sparse.csr_array` does not lead to this problem anymore but we need to be backward compatible with old scipy for some time still before being able to deprecate support for `csr_matrix`.

Reply:
```python
>>> import numpy as np
>>> from scipy.sparse import rand
>>> X = rand(100_000, 5)
>>> X
<100000x5 sparse matrix of type '<class 'numpy.float64'>'
        with 5000 stored elements in COOrdinate format>
>>> C = X.T @ X
>>> C
<5x5 sparse matrix of type '<class 'numpy.float64'>'
        with 25 stored elements in Compressed Sparse Row format>
>>> n_samples = X.shape[0]
>>> X_mean = X.mean(axis=0)
>>> X_mean
matrix([[0.00531109, 0.00497345, 0.00498098, 0.00488378, 0.00509504]])
>>> C -= (n_samples * X_mean.reshape(-1, 1) * X_mean.reshape(1, -1))
>>> C
matrix([[ 3.59159095e+02, -1.48415344e+00,  2.06023127e+00,
         -2.50133851e-02, -2.62356078e+00],
        [-1.48415344e+00,  3.33571655e+02,  7.52427579e-01,
          2.02376720e+00,  6.44612078e-01],
        [ 2.06023127e+00,  7.52427579e-01,  3.27488715e+02,
          2.18887774e+00, -8.94077160e-01],
        [-2.50133851e-02,  2.02376720e+00,  2.18887774e+00,
          3.24096148e+02,  8.62317950e-01],
        [-2.62356078e+00,  6.44612078e-01, -8.94077160e-01,
          8.62317950e-01,  3.37911475e+02]])
>>> C /= n_samples - 1
>>> eigenvals, eigenvecs = np.linalg.eigh(C)
>>> eigenvals
array([0.00322734, 0.00327965, 0.00334087, 0.00337746, 0.00359712])
>>> eigenvecs
matrix([[-0.01358159,  0.06717155, -0.023717  ,  0.13244301, -0.98853421],
        [ 0.13346428,  0.27465896, -0.94081235,  0.13532164,  0.05753191],
        [ 0.41103831, -0.88445837, -0.2085449 , -0.03245091, -0.06509116],
        [-0.89918035, -0.36812082, -0.22385077,  0.07644526,  0.00295262],
        [ 0.0672684 , -0.04765382,  0.14390791,  0.97839152,  0.12346913]])
```

Reply:
How about calling `eigenvals = xp.asarray(eigenvals)` and a TODO note for sparse matrix cleaning (once that happens)?